blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+- If the commit exists in `lsif_nearest_uploads`, we simply use those visible uploads, otherwise
+- If the commit exists in `lsif_nearest_upload_links`, we use the visible uploads attached to the ancestor.
+
+#### Example
+
+We'll use the following commit graph again for our example. Here, commit `68acd3` defines upload #1, and `67e0bf` defines upload #2 (both with distinct indexing root directories).
+
+<div class="no-shadow">
+  <img src="https://sourcegraphstatic.com/blog/commit-graph-optimizations/graph5.png" alt="sample commit graph">
+</div>
+
+The `lsif_nearest_uploads` table associates a commit with its visible uploads, just as before. But now, the number of records in the table is much, much smaller. The commits present in this table satisfy one of the properties described above that make the commit non-trivial to recompute.
+
+| repository                    | commit              | upload_ids          |
+| ----------------------------- | ------------------- | ------------------- |
+| github.com/sourcegraph/sample | <code>4a8a33</code> | <code>[]</code>     |
+| github.com/sourcegraph/sample | <code>68acd3</code> | <code>[1]</code>    |
+| github.com/sourcegraph/sample | <code>e43f5b</code> | <code>[1]</code>    |
+| github.com/sourcegraph/sample | <code>67e0bf</code> | <code>[2]</code>    |
+| github.com/sourcegraph/sample | <code>599611</code> | <code>[1, 2]</code> |
+
+Luckily, some benefits compound one another here, and after we started ignoring traversing the graph in both directions, we can simplify these properties to only account for ancestor-direction traversals. Notably, a commit whose parent has multiple children (`7e0471`, for example) were necessary to store for the descendant-direction traversal, but no longer need to be stored. This further increases the number of trivially recomputable commits, saving even more space.
+
+The `lsif_nearest_uploads_links` table stores a _forwarding pointer_ to the ancestor that has the same set of visible uploads.
+
+| repository                    | commit              | ancestor_commit     |
+| ----------------------------- | ------------------- | ------------------- |
+| github.com/sourcegraph/sample | <code>91a565</code> | <code>68acd3</code> |
+| github.com/sourcegraph/sample | <code>7e0471</code> | <code>4a8a33</code> |
+| github.com/sourcegraph/sample | <code>52811d</code> | <code>7e0471</code> |
+| github.com/sourcegraph/sample | <code>7b1a18</code> | <code>599611</code> |
+| github.com/sourcegraph/sample | <code>dd8578</code> | <code>599611</code> |
+
+Note that for our instances with a large number of distinct indexing roots, this saves a **massive** amount of storage space. The majority of commits (> 80%) can link to an ancestor, which requires only referencing a fixed-size commit hash. The remaining minority of commits must explicitly list their visible uploads, of which there may be many thousands.
```

Comment:
I feel like this ends a bit abruptly. I'm not suggesting a "Conclusion" paragraph to end, but it might be helpful to close with a summary of what you did/learned?

Reply:
I left a few todos to myself to massage some of the content in part 2.